service/resource/vaultpki/create.go

```diff
@@ -0,0 +1,77 @@
+package vaultpki
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/cert-operator/service/key"
+)
+
+func (r *Resource) GetCreateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	currentVaultPKIState, err := toVaultPKIState(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	desiredVaultPKIState, err := toVaultPKIState(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.Log("cluster", key.ClusterID(customObject), "debug", "finding out if the Vault PKI has to be created")
+
+	var vaultPKIStateToCreate VaultPKIState
+	if currentVaultPKIState.Backend == nil {
+		vaultPKIStateToCreate.Backend = desiredVaultPKIState.Backend
+	}
+	if currentVaultPKIState.CACertificate == "" {
+		vaultPKIStateToCreate.CACertificate = desiredVaultPKIState.CACertificate
+	}
+
+	r.logger.Log("cluster", key.ClusterID(customObject), "debug", "found out if the Vault PKI has to be created")
+
+	return vaultPKIStateToCreate, nil
+}
+
+func (r *Resource) ProcessCreateState(ctx context.Context, obj, createState interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	vaultPKIStateToCreate, err := toVaultPKIState(createState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if vaultPKIStateToCreate.Backend != nil {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "creating the Vault PKI in the Vault API")
+
+		err := r.vaultPKI.CreateBackend(key.ClusterID(customObject))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "created the Vault PKI in the Vault API")
+	} else {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "the Vault PKI does not need to be created in the Vault API")
+	}
+
+	if vaultPKIStateToCreate.CACertificate != "" {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "creating the root CA in the Vault PKI")
+
+		err := r.vaultPKI.CreateCA(key.ClusterID(customObject))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "created the root CA in the Vault PKI")
```

Comment:
this reads oddly to me - it's the vault pki backend, right?

Reply:
Yes. Would it be better to add the word `backend` here?

Reply:
Not fussed. Do it if you care, I get it now :D